File: frontend/app/src/lib/player-room.ts
import { useCallback, useEffect, useRef, useState } from 'react'

import type { components } from '@/api/schema'

export type PlayerRoomMode = 'sheet' | 'av' | 'slide'
export type PlayerRoomSourceType = 'song' | 'collection' | 'setlist'
export type PlayerRoomMusicalState = { item_index: number; language: string | null; transposition: string | null }
export type PlayerRoomProjection = {
  content_text: string
  content_lines?: unknown
  content_layer: Record<string, unknown>
  background_layer: Record<string, unknown>
  transition: Record<string, unknown>
  screen_state: 'live' | 'blank' | 'blackout'
  item_title: string
  next_preview: string | null
}
export type PlayerRoomParticipant = { id: string; mode: PlayerRoomMode; hide_chords?: boolean; display_name: string; avatar_url: string | null; anonymous: boolean; connected: boolean; is_host: boolean; is_av_host: boolean }
export type PlayerRoomSummary = { id: string; name: string; team_id: string; source_type: PlayerRoomSourceType; source_id: string; source_title: string; host_email: string; participant_count: number; av_occupied: boolean; created_at: string }
export type PlayerRoomSnapshot = PlayerRoomSummary & { content: { items: components['schemas']['Player']['items']; toc: components['schemas']['Player']['toc'] }; musical_state: PlayerRoomMusicalState; projection: PlayerRoomProjection | null; participants: PlayerRoomParticipant[]; revision: number; host_lease_expires_at: string; guests_allowed?: boolean }
export type PlayerRoomCredentials = { room_id: string; participant_id: string; mode: PlayerRoomMode; resume_credential: string; connection_ticket: string }
export type CreatedPlayerRoom = { room: PlayerRoomSummary; credentials: PlayerRoomCredentials; invite_secret: string }

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? '').replace(/\/$/, '')
const credentialKey = (roomId: string) => `playerRoom:${roomId}:credentials`
const inviteKey = (roomId: string) => `playerRoom:${roomId}:invite`
const REDACTED_ROOM_EVENT_VALUE = '[redacted]'

export function redactPlayerRoomEvent(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(redactPlayerRoomEvent)
  if (!value || typeof value !== 'object') return value
  return Object.fromEntries(
    Object.entries(value).map(([key, nested]) => [
      key,
      /ticket|credential|secret/i.test(key)
        ? REDACTED_ROOM_EVENT_VALUE
        : redactPlayerRoomEvent(nested),
    ]),
  )
}

function logPlayerRoomEvent(
  roomId: string,
  direction: 'incoming' | 'outgoing',
  event: unknown,
): void {
  console.log(`[PlayerRoom ${roomId}] ${direction}`, redactPlayerRoomEvent(event))
}

function sendPlayerRoomEvent(roomId: string, socket: WebSocket, event: object): void {
  logPlayerRoomEvent(roomId, 'outgoing', event)
  socket.send(JSON.stringify(event))
}

async function jsonRequest<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBase}${path}`, { credentials: 'include', ...init, headers: { 'Content-Type': 'application/json', ...init?.headers } })
  if (!response.ok) {
    const problem = await response.clone().json().catch(async () => ({
      detail: await response.text().catch(() => ''),
    }))
    console.warn('[PlayerRoom HTTP] request failed', {
      method: init?.method ?? 'GET',
      path,
      status: response.status,
      problem: redactPlayerRoomEvent(problem),
    })
    const code = problem && typeof problem === 'object' && 'code' in problem
      ? String(problem.code)
      : response.status === 409 ? 'av_occupied' : 'player_room_unavailable'
    throw new Error(code)
  }
  return response.json() as Promise<T>
}

export function saveRoomCredentials(credentials: PlayerRoomCredentials): void { sessionStorage.setItem(credentialKey(credentials.room_id), JSON.stringify(credentials)) }
export function readRoomCredentials(roomId: string): PlayerRoomCredentials | null { try { const raw = sessionStorage.getItem(credentialKey(roomId)); return raw ? JSON.parse(raw) as PlayerRoomCredentials : null } catch { return null } }
export function saveRoomInvite(roomId: string, secret: string): void { sessionStorage.setItem(inviteKey(roomId), secret) }
export function readRoomInvite(roomId: string): string | null { return sessionStorage.getItem(inviteKey(roomId)) }

export async function createPlayerRoom(input: { source_type: PlayerRoomSourceType; source_id: string; host_mode: Exclude<PlayerRoomMode, 'slide'>; musical_state: PlayerRoomMusicalState; projection: PlayerRoomProjection | null }): Promise<CreatedPlayerRoom> {
  console.log('[PlayerRoom HTTP] create', redactPlayerRoomEvent(input))
  const created = await jsonRequest<CreatedPlayerRoom>('/api/v1/player-rooms', { method: 'POST', body: JSON.stringify(input) })
  saveRoomCredentials(created.credentials); saveRoomInvite(created.room.id, created.invite_secret); return created
}
export async function listPlayerRooms(params: { page: number; q?: string; team?: string }): Promise<{ items: PlayerRoomSummary[]; total: number }> {
  const search = new URLSearchParams({ page: String(params.page), page_size: '50' }); if (params.q) search.set('q', params.q); if (params.team) search.set('team', params.team)
  const response = await fetch(`${apiBase}/api/v1/player-rooms?${search}`, { credentials: 'include' }); if (!response.ok) throw new Error('player_room_unavailable')
  return { items: await response.json() as PlayerRoomSummary[], total: Number(response.headers.get('x-total-count') ?? 0) }
}
export async function joinPlayerRoom(roomId: string, mode: PlayerRoomMode, hideChords = false): Promise<PlayerRoomCredentials> {
  const previous = readRoomCredentials(roomId); const credentials = await jsonRequest<PlayerRoomCredentials>(`/api/v1/player-rooms/${encodeURIComponent(roomId)}/join`, { method: 'POST', body: JSON.stringify({ mode, hide_chords: hideChords, resume_credential: previous?.mode === mode ? previous.resume_credential : null }) }); saveRoomCredentials(credentials); return credentials
}
export async function inspectPlayerRoomInvite(inviteSecret: string): Promise<{ room_id: string; name: string; host_email: string; av_occupied: boolean; guests_allowed?: boolean }> { return jsonRequest('/api/v1/player-rooms/invite/inspect', { method: 'POST', body: JSON.stringify({ invite_secret: inviteSecret }) }) }
export async function joinPlayerRoomInvite(input: { invite_secret: string; display_name: string; mode: PlayerRoomMode; hide_chords?: boolean }): Promise<PlayerRoomCredentials> { const credentials = await jsonRequest<PlayerRoomCredentials>('/api/v1/player-rooms/invite/join', { method: 'POST', body: JSON.stringify({ hide_chords: false, ...input }) }); saveRoomCredentials(credentials); return credentials }
export async function endPlayerRoom(roomId: string, resumeCredential: string): Promise<void> { const response = await fetch(`${apiBase}/api/v1/player-rooms/${encodeURIComponent(roomId)}`, { method: 'DELETE', credentials: 'include', headers: { 'X-Player-Room-Credential': resumeCredential } }); if (!response.ok) throw new Error('player_room_unavailable') }
async function reconnectPlayerRoom(credentials: PlayerRoomCredentials): Promise<PlayerRoomCredentials> { const next = await jsonRequest<PlayerRoomCredentials>(`/api/v1/player-rooms/${encodeURIComponent(credentials.room_id)}/reconnect`, { method: 'POST', body: JSON.stringify({ mode: credentials.mode, resume_credential: credentials.resume_credential }) }); saveRoomCredentials(next); return next }

export type RoomConnection = { snapshot: PlayerRoomSnapshot | null; status: 'connecting' | 'connected' | 'reconnecting' | 'ended'; sendMusicalState: (state: PlayerRoomMusicalState) => void; sendProjection: (projection: PlayerRoomProjection) => void; sendGuestsAllowed: (guestsAllowed: boolean) => void; leave: () => void }

export function usePlayerRoom(credentials: PlayerRoomCredentials | null): RoomConnection {
  const [snapshot, setSnapshot] = useState<PlayerRoomSnapshot | null>(null)
  const [status, setStatus] = useState<RoomConnection['status']>('connecting')
  const socketRef = useRef<WebSocket | null>(null); const retryRef = useRef(0); const closedRef = useRef(false)
  const send = useCallback((message: object) => {
    if (credentials && socketRef.current?.readyState === WebSocket.OPEN) {
      sendPlayerRoomEvent(credentials.room_id, socketRef.current, message)
    }
  }, [credentials])
  useEffect(() => {
    if (!credentials) return
    closedRef.current = false; let disposed = false; let retryTimer: number | undefined; let heartbeat: number | undefined
    const connect = async () => {
      setStatus(retryRef.current ? 'reconnecting' : 'connecting')
      console.log(`[PlayerRoom ${credentials.room_id}] connecting`, { attempt: retryRef.current + 1 })
      let activeCredentials = credentials
      if (retryRef.current > 0) {
        const reconnected = await reconnectPlayerRoom(credentials).catch(() => null)
        if (disposed) return
        if (!reconnected) {
          console.warn(`[PlayerRoom ${credentials.room_id}] credential exchange failed`)
          setStatus('reconnecting')
          const delay = Math.min(10_000, 500 * 2 ** retryRef.current++)
          console.log(`[PlayerRoom ${credentials.room_id}] reconnect scheduled`, { delayMs: delay })
          retryTimer = window.setTimeout(connect, delay)
          return
        }
        activeCredentials = reconnected
      }
      const base = apiBase || window.location.origin; const url = new URL('/api/v1/player-rooms/ws', base); url.protocol = url.protocol === 'https:' ? 'wss:' : 'ws:'
      const ws = new WebSocket(url); socketRef.current = ws
      ws.onopen = () => {
        console.log(`[PlayerRoom ${credentials.room_id}] socket open`)
        sendPlayerRoomEvent(credentials.room_id, ws, { type: 'authenticate', ticket: activeCredentials.connection_ticket })
        heartbeat = window.setInterval(() => {
          if (ws.readyState === WebSocket.OPEN) sendPlayerRoomEvent(credentials.room_id, ws, { type: 'heartbeat' })
        }, 10_000)
      }
      ws.onmessage = (event) => {
        let message: { type: string; snapshot?: PlayerRoomSnapshot }
        try {
          message = JSON.parse(String(event.data)) as typeof message
        } catch (error) {
          console.warn(`[PlayerRoom ${credentials.room_id}] invalid incoming event`, {
            dataType: typeof event.data,
            dataLength: typeof event.data === 'string' ? event.data.length : undefined,
            error,
          })
          return
        }
        logPlayerRoomEvent(credentials.room_id, 'incoming', message)
        if (message.type === 'room_ended') { closedRef.current = true; setStatus('ended'); ws.close(); return }
        if (message.snapshot) { setSnapshot((current) => { if (current && message.type === 'state_updated' && message.snapshot!.revision > current.revision + 1) { sendPlayerRoomEvent(credentials.room_id, ws, { type: 'request_snapshot' }); return current } return !current || message.snapshot!.revision >= current.revision ? message.snapshot! : current }); setStatus('connected'); retryRef.current = 0 }
      }
      ws.onerror = () => console.warn(`[PlayerRoom ${credentials.room_id}] socket error`)
      ws.onclose = (event) => {
        console.log(`[PlayerRoom ${credentials.room_id}] socket closed`, { code: event.code, reason: event.reason, wasClean: event.wasClean })
        if (heartbeat) window.clearInterval(heartbeat)
        if (closedRef.current) return
        setStatus('reconnecting')
        const delay = Math.min(10_000, 500 * 2 ** retryRef.current++)
        console.log(`[PlayerRoom ${credentials.room_id}] reconnect scheduled`, { delayMs: delay })
        retryTimer = window.setTimeout(connect, delay)
      }
    }
    void connect(); return () => { disposed = true; closedRef.current = true; if (retryTimer) window.clearTimeout(retryTimer); if (heartbeat) window.clearInterval(heartbeat); socketRef.current?.close() }
  }, [credentials])
  const sendMusicalState = useCallback(
    (musical_state: PlayerRoomMusicalState) =>
      send({ type: 'update_musical_state', command_id: crypto.randomUUID(), musical_state }),
    [send],
  )
  const sendProjection = useCallback(
    (projection: PlayerRoomProjection) =>
      send({ type: 'update_projection', command_id: crypto.randomUUID(), projection }),
    [send],
  )
  const sendGuestsAllowed = useCallback(
    (guests_allowed: boolean) =>
      send({ type: 'update_guests_allowed', command_id: crypto.randomUUID(), guests_allowed }),
    [send],
  )
  const leave = useCallback(() => {
    send({ type: 'leave' })
    closedRef.current = true
    socketRef.current?.close()
  }, [send])
  return { snapshot, status, sendMusicalState, sendProjection, sendGuestsAllowed, leave }
}

export function playerFromRoom(snapshot: PlayerRoomSnapshot): components['schemas']['Player'] { return { items: snapshot.content.items, toc: snapshot.content.toc, scroll_type: 'one_page', scroll_type_cache_other_orientation: 'book', orientation: 'portrait', between_items: false, index: snapshot.musical_state.item_index } }

export function participantModeLabel(
  participant: Pick<PlayerRoomParticipant, 'mode' | 'hide_chords'>,
  t: (key: string) => string,
): string {
  if (participant.mode === 'sheet') {
    return t(participant.hide_chords ? 'playerRooms.mode.text' : 'playerRooms.mode.chords')
  }
  return t(`playerRooms.mode.${participant.mode}`)
}

export function formatRoomDuration(totalSeconds: number): string {
  const hours = Math.floor(totalSeconds / 3600)
  const minutes = Math.floor((totalSeconds % 3600) / 60)
  const seconds = totalSeconds % 60
  if (hours > 0) {
    return `${hours}:${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`
  }
  return `${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`
}

export function useRoomElapsedSeconds(since: string): number {
  const [elapsed, setElapsed] = useState(() =>
    Math.max(0, Math.floor((Date.now() - Date.parse(since)) / 1000)),
  )

  useEffect(() => {
    const tick = () => {
      setElapsed(Math.max(0, Math.floor((Date.now() - Date.parse(since)) / 1000)))
    }
    tick()
    const id = window.setInterval(tick, 1000)
    return () => window.clearInterval(id)
  }, [since])

  return elapsed
}

export function roomSourceTypeLabel(
  sourceType: PlayerRoomSourceType,
  t: (key: string) => string,
): string {
  return t(`playerRooms.sourceType.${sourceType}`)
}

export function playerRoomShortName(room: Pick<PlayerRoomSummary, 'source_title'>): string {
  return room.source_title
}

File: shared/src/player_room.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
#[cfg(feature = "backend")]
use utoipa::ToSchema;

use crate::player::{Player, PlayerChordsItem, PlayerItem, TocItem};

#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub enum PlayerRoomSourceType {
    Song,
    Collection,
    Setlist,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(tag = "type", rename_all = "snake_case")]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub enum PlayerRoomSongPool {
    Open,
    Collection { id: String, title: String },
    Setlist { id: String, title: String },
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(tag = "type", rename_all = "snake_case")]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub enum PlayerRoomSongPoolSelection {
    Open,
    Collection { id: String },
    Setlist { id: String },
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct UpdatePlayerRoomSongPool {
    pub pool: PlayerRoomSongPoolSelection,
    pub revision: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub enum PlayerRoomMode {
    Sheet,
    Av,
    Slide,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomContent {
    pub items: Vec<PlayerItem>,
    pub toc: Vec<TocItem>,
}

impl From<&Player> for PlayerRoomContent {
    fn from(player: &Player) -> Self {
        let mut items = player.items().to_vec();
        for item in &mut items {
            if let PlayerItem::Chords(chords) = item {
                chords.song.user_specific_addons.liked = false;
            }
        }
        let toc = player
            .toc()
            .iter()
            .cloned()
            .map(|mut row| {
                row.liked = false;
                row
            })
            .collect();
        Self { items, toc }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Default)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomMusicalState {
    pub item_index: usize,
    pub language: Option<String>,
    pub transposition: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Default)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomProjectionPayload {
    pub content_text: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "backend", schema(value_type = Option<Object>))]
    pub content_lines: Option<serde_json::Value>,
    #[cfg_attr(feature = "backend", schema(value_type = Object))]
    pub content_layer: serde_json::Value,
    #[cfg_attr(feature = "backend", schema(value_type = Object))]
    pub background_layer: serde_json::Value,
    #[cfg_attr(feature = "backend", schema(value_type = Object))]
    pub transition: serde_json::Value,
    pub screen_state: String,
    pub item_title: String,
    pub next_preview: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomQueueItem {
    pub id: String,
    pub song_id: String,
    pub title: String,
    pub song: Box<PlayerChordsItem>,
    pub added_by: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomParticipant {
    pub id: String,
    pub mode: PlayerRoomMode,
    #[serde(default)]
    pub hide_chords: bool,
    pub display_name: String,
    pub avatar_url: Option<String>,
    pub anonymous: bool,
    pub connected: bool,
    pub is_host: bool,
    pub is_av_host: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomSummary {
    pub id: String,
    pub name: String,
    pub team_id: String,
    pub source_type: Option<PlayerRoomSourceType>,
    pub source_id: Option<String>,
    pub source_title: Option<String>,
    #[serde(default = "default_song_pool")]
    pub song_pool: PlayerRoomSongPool,
    pub host_email: String,
    #[serde(default)]
    pub can_close: bool,
    pub participant_count: usize,
    pub av_occupied: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomSnapshot {
    #[serde(flatten)]
    pub summary: PlayerRoomSummary,
    pub content: PlayerRoomContent,
    #[serde(default)]
    pub queue: Vec<PlayerRoomQueueItem>,
    pub musical_state: PlayerRoomMusicalState,
    pub projection: Option<PlayerRoomProjectionPayload>,
    pub participants: Vec<PlayerRoomParticipant>,
    pub revision: u64,
    pub host_lease_expires_at: DateTime<Utc>,
    #[serde(default = "default_guests_allowed")]
    pub guests_allowed: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct AddPlayerRoomQueueItem {
    pub song_id: String,
    pub revision: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct ReorderPlayerRoomQueue {
    pub queue_ids: Vec<String>,
    pub revision: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomQueueRevision {
    pub revision: u64,
}

fn default_guests_allowed() -> bool {
    true
}

fn default_song_pool() -> PlayerRoomSongPool {
    PlayerRoomSongPool::Open
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct CreatePlayerRoom {
    pub team_id: String,
    #[serde(default)]
    pub name: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct JoinPlayerRoom {
    pub mode: PlayerRoomMode,
    #[serde(default)]
    pub hide_chords: bool,
    pub resume_credential: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct InspectPlayerRoomInvite {
    pub invite_secret: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct JoinPlayerRoomInvite {
    pub invite_secret: String,
    pub display_name: String,
    pub mode: PlayerRoomMode,
    #[serde(default)]
    pub hide_chords: bool,
    pub resume_credential: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomInviteInfo {
    pub room_id: String,
    pub name: String,
    pub host_email: String,
    pub av_occupied: bool,
    #[serde(default = "default_guests_allowed")]
    pub guests_allowed: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct PlayerRoomCredentials {
    pub room_id: String,
    pub participant_id: String,
    pub mode: PlayerRoomMode,
    pub resume_credential: String,
    pub connection_ticket: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[cfg_attr(feature = "backend", derive(ToSchema))]
pub struct CreatedPlayerRoom {
    pub room: PlayerRoomSummary,
    pub credentials: PlayerRoomCredentials,
    pub invite_secret: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        player::{Player, PlayerChordsItem, PlayerItem},
        song::{Song, SongUserSpecificAddons},
    };

    #[test]
    fn room_content_strips_user_likes() {
        let mut song = Song {
            id: "song-1".into(),
            user_specific_addons: SongUserSpecificAddons { liked: true },
            ..Default::default()
        };
        song.data.titles = vec!["Title".into()];
        let player = Player::new(
            vec![PlayerItem::Chords(Box::new(PlayerChordsItem {
                song,
                language: None,
                flow: None,
            }))],
            vec![TocItem {
                idx: 0,
                title: "Title".into(),
                id: Some("song-1".into()),
                nr: String::new(),
                liked: true,
            }],
        );
        let content = PlayerRoomContent::from(&player);
        assert!(!content.toc[0].liked);
        let PlayerItem::Chords(item) = &content.items[0] else {
            panic!("expected chords")
        };
        assert!(!item.song.user_specific_addons.liked);
    }

    #[test]
    fn song_pool_values_roundtrip_with_tagged_json() {
        let pool = PlayerRoomSongPool::Collection {
            id: "collection-1".into(),
            title: "Sunday songs".into(),
        };
        let json = serde_json::to_string(&pool).expect("song pool JSON");
        assert_eq!(
            json,
            r#"{"type":"collection","id":"collection-1","title":"Sunday songs"}"#
        );
        assert_eq!(
            serde_json::from_str::<PlayerRoomSongPool>(&json).unwrap(),
            pool
        );

        let selection = PlayerRoomSongPoolSelection::Setlist {
            id: "setlist-1".into(),
        };
        assert_eq!(
            serde_json::from_str::<PlayerRoomSongPoolSelection>(
                r#"{"type":"setlist","id":"setlist-1"}"#
            )
            .unwrap(),
            selection
        );
    }

    #[test]
    fn older_room_summaries_default_to_open_song_pool() {
        let json = r#"{
            "id":"room-1","name":"Room","team_id":"team-1",
            "source_type":null,"source_id":null,"source_title":null,
            "host_email":"host@example.com","can_close":false,
            "participant_count":0,"av_occupied":false,
            "created_at":"2026-01-01T00:00:00Z"
        }"#;
        let summary: PlayerRoomSummary = serde_json::from_str(json).unwrap();
        assert_eq!(summary.song_pool, PlayerRoomSongPool::Open);
    }
}
